src/main/java/org/seqdoop/hadoop_bam/BAMFilteredRecordReader.java

```diff
@@ -0,0 +1,151 @@
+    
+    @Override public void initialize(InputSplit spl,
+                                     TaskAttemptContext ctx) 
+        throws IOException {
+        // Check to ensure this method is only be called once (see Hadoop API)
+        if (isInitialized) {
+            close();
+        }
+        isInitialized = true;
+
+        Configuration conf = ContextUtil.getConfiguration(ctx);
+        FileVirtualSplit split = (FileVirtualSplit) spl;
+        Path file = split.getPath();
+        FileSystem fs = file.getFileSystem(conf);
+
+        this.stringency = SAMHeaderReader.getValidationStringency(conf);
+
+        FSDataInputStream in = fs.open(file);
+
+        // Sets codec to translate between in-memory and disk representation of record
+        codec = new BAMRecordCodec(SAMHeaderReader.readSAMHeaderFrom(in, conf));
+        
+        in.seek(0);
+
+        bci = new BlockCompressedInputStream(new WrapSeekable<FSDataInputStream>(in,
+                                                                                 fs.getFileStatus(file).getLen(),
+                                                                                 file));
+
+        // Gets BGZF virtual offset for the split
+        long virtualStart = split.getStartVirtualOffset();
+
+        fileStart = virtualStart >>> 16;
+        virtualEnd = split.getEndVirtualOffset();
+
+        // Starts looking from the BGZF virtual offset
+        bci.seek(virtualStart);
+        // Reads records from this input stream
+        codec.setInputStream(bci);
+    }
+    
+    @Override public void close() throws IOException {
+        bci.close();
+    }
+
+    @Override public LongWritable getCurrentKey() {
+        return key;
+    }
+
+    @Override public SAMRecordWritable getCurrentValue() {
+        return record;
+    }
+
+    /**
+     * This method gets the nextKeyValue for our RecordReader, but filters by only
+     * returning records within a specified ReferenceRegion.
+     */
+    @Override public boolean nextKeyValue() {
+      while (bci.getFilePointer() < virtualEnd) {
+          SAMRecord r = codec.decode();
+
+          // Since we're reading from a BAMRecordCodec directly we have to set the
+          // validation stringency ourselves.
+          if (this.stringency != null) {
+              r.setValidationStringency(this.stringency);
+          }
+
+          // This if/else block pushes the predicate down onto a BGZIP block that 
+          // the index has said contains data in our specified region.
+          if (r == null) {
+              return false;
+          } else {
+              int start = r.getStart();
+              int end = r.getEnd();
+
+              if ((r.getContig() == viewContig) &&
+                  (((start >= viewStart) && (end <= viewEnd))
+                   || ((start <= viewStart) && (end >= viewStart) && (end <= viewEnd))
+                   || ((end >= viewEnd) && (start >= viewStart) && (start <= viewEnd)))) {
```

Comment:
I think this condition is overly restrictive, as it doesn't cover the case where the interval (viewStart, viewEnd) is wholly contained in (start, end). I tried changing the condition to `htsjdk.samtools.util.CoordMath.overlaps(start, end, viewStart, viewEnd)` and it worked for the testing I've been doing for https://github.com/broadinstitute/gatk/issues/1399.

Unmapped reads with a position need to be taken care of too - see https://github.com/broadinstitute/gatk/blob/master/src/main/java/org/broadinstitute/hellbender/engine/spark/datasources/ReadsSparkSource.java#L210